lib/index.ts

```diff
@@ -0,0 +1,292 @@
+        this.entries[toIndex].relativePath === entries[fromIndex].relativePath) {
+        this.entries.splice(toIndex, 1, entries[fromIndex++]);
+      } else {
+        this.entries.splice(toIndex++, 0, entries[fromIndex++]);
+      }
+    }
+  }
+
+  addPaths(paths: string[], options?: FSTree.StaticOptions) {
+    const entries = paths.map(path => {
+      return new Entry(path, 0, ARBITRARY_START_OF_TIME);
+    });
+
+    this.addEntries(entries, options);
+  }
+
+  forEach(fn: (entry: Entry, index: number, collection: Entry[]) => void, context: any) {
+    this.entries.forEach(fn, context);
+  }
+
+  calculatePatch(otherFSTree: FSTree, isEqual?: (a: IEntry, b: IEntry) => boolean) : FSTree.Patch {
+    if (arguments.length > 1 && typeof isEqual !== 'function') {
+      throw new TypeError('calculatePatch\'s second argument must be a function');
+    }
+
+    if (typeof isEqual !== 'function') {
+      isEqual = (this.constructor as typeof FSTree).defaultIsEqual;
+    }
+
+    const ours = this.entries;
+    const theirs = otherFSTree.entries;
+    const additions = [];
+
+    let i = 0;
+    let j = 0;
+
+    const removals: FSTree.Patch = [];
+
+    let command;
+
+    while (i < ours.length && j < theirs.length) {
+      let x = ours[i];
+      let y = theirs[j];
+
+      if (x.relativePath < y.relativePath) {
+        // ours
+        i++;
+
+        removals.push(removeOperation(x));
+
+        // remove additions
+      } else if (x.relativePath > y.relativePath) {
+        // theirs
+        j++;
+        additions.push(addOperation(y));
+      } else {
+        if (!isEqual(x, y)) {
+          command = updateOperation(y);
+
+          if (x.isDirectory()) {
+            removals.push(command);
+          } else {
+            additions.push(command);
+          }
+        }
+        // both are the same
+        i++; j++;
+      }
+    }
+
+    // cleanup ours
+    for (; i < ours.length; i++) {
+      removals.push(removeOperation(ours[i]));
+    }
+
+    // cleanup theirs
+    for (; j < theirs.length; j++) {
+      additions.push(addOperation(theirs[j]));
+    }
+
+    // operations = removals (in reverse) then additions
+    return removals.reverse().concat(additions);
+  }
+
+  calculateAndApplyPatch(otherFSTree: FSTree, input: string, output: string, delegate?: FSTree.PatchDelegate) {
+    (this.constructor as typeof FSTree).applyPatch(input, output, this.calculatePatch(otherFSTree), delegate);
+  }
+
+  static defaultIsEqual(entryA: FSTree.IEntry, entryB: FSTree.IEntry) {
+    if (entryA.isDirectory() && entryB.isDirectory()) {
+      // ignore directory changes by default
+      return true;
+    }
+
+    let equal;
+    if (entryA.size === entryB.size && entryA.mode === entryB.mode) {
+      if (entryA.mtime === entryB.mtime) {
+        equal = true;
+      } else if (entryA.mtime === undefined || entryB.mtime === undefined) {
+        equal = false;
+      } else if (+entryA.mtime === +entryB.mtime) {
+        equal = true;
```

Comment:
I'd think you'd want to early `return` in these `equal = true;` cases.

Reply:
Ya, i toyed with several variations. The current ends up with the one i enjoyed the most.